### ISSUE TITLE
Can move priors become so biased to prevent learning even with noise?

### ISSUE THREAD

**Issue by Mardak** [leela-zero/leela-zero#1588]
In analyzing Leela Chess Zero behavior of not learning some tactics https://github.com/LeelaChessZero/lc0/issues/8, I noticed that root search using only value without policy would easily find the correct move, which usually had low 0.1% priors, because the tactical move had a favorable win rate to get more visits from search.

This got me wondering if Go would run into similar issues, so I forced 1000 visits into the top 3 moves from this board state:
```
   a b c d e f g h j k l m n o p q r s t 
19 . . . . . . . . . . . . . . . . . . . 19
18 . . . . . . . . . . . . . . . . . . . 18
17 . . . . . . . . . . . . . O . . . . . 17
16 . . . X . . . . . + . . . . . O . . . 16
15 . . . . . . . . . . . . . . . . . . . 15
14 . . . . . . . . . . . . . . . . X . . 14
13 . . . . . . . . . . . . . . . . . . . 13
12 . . . . . . . . . . . . . . . . . . . 12
11 . . . . . . . . . . . . . . . . . . . 11
10 . . . + . . . . . + . . . . . + . . . 10
 9 . . . . . . . . . . . . . . . . . . .  9
 8 . . . . . . . . . . . . . . . . . . .  8
 7 . . . . . . . . . . . . . . . . . . .  7
 6 . . . . . . . . . . . . . . . . . . .  6
 5 . . . . . . . . . . . . . . . . . . .  5
 4 . . O O . . . . . + . . . . . X . . .  4
 3 . . X(X). . . . . . . . . . . . . . .  3
 2 . . . . . . . . . . . . . . . . . . .  2
 1 . . . . . . . . . . . . . . . . . . .  1
   a b c d e f g h j k l m n o p q r s t 
```

Here's the results with the latest networks:
```
LZ150 (2b80a9db)
  F3 ->    1000 (V: 53.05%) (N: 64.10%) PV: F3 E3 E4 F2 R3 Q3 R4 R6 R5 Q5 S6 G3 F4 B3 H5 R2 S2 S7 Q2 Q7 P6 Q6 P2 F17
  E4 ->    1000 (V: 52.87%) (N:  0.81%) PV: E4 F2 R3 Q3 R4 R6 R5 Q5 S6 G3 B3 B2 D9 C11 C17 D17 C16
  E3 ->    1000 (V: 51.98%) (N: 32.51%) PV: E3 E2 F2 F3 E4 G2 D2 F1 C2 R16 R17 S17 R15 S16 Q15 S15 C17 D17 C16 C14 C15 D15 B14 Q14 L17 C18 B18 B13 D18 D13

256V16 (c010f034)
  F3 ->    1000 (V: 53.31%) (N: 62.62%) PV: F3 E3 E4 F2 R3 Q3 R4 R6 R5 Q5 S6 G3 F4 B3 H5 R2 S2 S7 Q6 R7 Q2 P2 R1 P6 C17 C16 D17
  E4 ->    1000 (V: 53.30%) (N:  0.81%) PV: E4 F2 R3 Q3 R4 R6 R5 Q5 S6 G3 B3 B2 C8 R2 S2 S5 Q2 S3 R1 S4 R2
  E3 ->    1000 (V: 52.46%) (N: 33.97%) PV: E3 E2 F2 F3 E4 G2 D2 F1 C2 R16 R17 S17 R15 S16 Q15 S15 R3 B3 B2 Q3 R4 R6 R5 Q5 S6 Q14 R7

ELF (62b5417b)
  E3 ->    1000 (V: 51.80%) (N: 28.58%) PV: E3 E2 F2 F3 E4 G2 D2 F1 C2 B3 B2 O3 C17 C16 D17 F17 E17 E16 F18 F4 G17 F16 B17 N7 G16 G15 H15 G14 H14 H13
  F3 ->    1000 (V: 51.21%) (N: 69.09%) PV: F3 E3 E4 F2 R3 Q3 R4 Q5 R5 R6 S6 G3 F4 B3 H5 D7 F7 D10 R15 Q14 R7
  E4 ->    1000 (V: 49.42%) (N:  2.31%) PV: E4 F2 R3 Q3 R4 R6 R5 Q5 S6 C8 F3 G2 H4 R2 S2 S5 S4 S7 Q6 R7
```

The 192x15 and 256x20 are pretty similar in behavior as they use the same training data and the board state isn't that complex. Notably, with equal visits, the 3rd-highest prior move E4's win rate is higher than the 2nd-rank move E3, so search would give it a lot more visits if the prior wasn't so low. In some sense, policy is just a reflection of value-maximizing behavior of training search data so that at match time, search isn't wasted on low value moves. However, looking at ELF, it believes E4 is actually not that great of a move, and leelaz would eventually get there with noise and T=1 playing E4 in training to get a more accurate win rate.

The search output for ELF by itself is also interesting because its 2nd-highest-prior move E3 has a higher win rate than F3. But the higher win rate isn't reflected in the policy where all 3 networks think F3 is the most likely move with a prior nearly double the next best move. Does this mean continued ELF training would have eventually prioritized E3 over F3? (Playing E3 here is actually the "flower joseki" which is by far the most common joseki in leelaz history http://zero.sjeng.org/opening.)

Looking at the next move assuming white played F3:
```
LZ150
  F2 ->    1000 (V: 47.39%) (N: 26.11%) PV: F2 G2 E2 G4 E4 E3 B4 B5 B3 C6 E5 D6 E6 E7 F7 F6 E8 D7 G6 F5 G5 F4 H4 G3 G7 R3 R4 Q3 P4 O2 R16 R17 S17 R15 S16 Q15 S15 R18 Q14
  E3 ->    1000 (V: 46.70%) (N: 71.04%) PV: E3 E4 F2 R3 Q3 R4 R6 R5 Q5 S6 G3 F4 B3 H5 R2 S2 S7 Q2 Q7 P6 Q6 P2 F17 C14
  E4 ->    1000 (V: 39.31%) (N:  0.03%) PV: E4 E3 B4 B5 B3 C6 E5 D6 G4 G3 J4 D2 C2 E2 H3 F4 F5 G5 H4 G6 G2 F6 E1 E6

256V16
  E3 ->    1000 (V: 46.67%) (N: 68.31%) PV: E3 E4 F2 R3 Q3 R4 R6 R5 Q5 S6 G3 F4 B3 H5 R2 S2 S7 Q6 R7 Q2 P2 R1 P6 C17 C16 D17 F17 E17
  F2 ->    1000 (V: 46.65%) (N: 28.44%) PV: F2 G2 E2 G4 E4 E3 B4 B5 B3 C6 E5 D6 E6 E7 F7 F6 E8 D7 G6 F5 G5 F4 H4 G3 G7 R3 R4 Q3 O3 P3 P4 O2 R16 R17 S17 R15 S16 Q15 S15 R18 Q14 S13 S14 R9 R11 P9 M17 O15
  B4 ->    1000 (V: 42.75%) (N:  0.03%) PV: B4 B5 E3 E4 F2 B3 B2 A4 G3 F4 H2 R3 Q3 R4 R6 R5 Q5 S6 R2 S2 S5 Q2 S3 R1 S4 R2 S7 Q6 R7 O3 O6 C17 C16 D17 F17 E17 E16 F18 P14 S16

ELF
  F2 ->    1000 (V: 49.65%) (N: 22.23%) PV: F2 G2 E2 G4 E4 E3 B4 B5 B3 C6 E5 D6 E6 E7 G5 F1 D2 H5 F7 E8 H6 J5 J6 K5 K6 L5 N3 L3 C8 C9
  E3 ->    1000 (V: 48.32%) (N: 77.73%) PV: E3 E4 F2 R3 Q3 R4 Q5 R5 R6 S6 G3 F4 B3 H5 D7 F7 D10 R15 Q14 R7 Q6
  B4 ->    1000 (V: 39.14%) (N:  0.01%) PV: B4 B5 E3 E4 F2 B3 B2 A4 G3 F4 H2 R3 Q3 R4 R6 R5 Q5 S6 R7 S7 R8 C17 C16 D17 F17 E17 E16 F18 H5 D11 B17 B18
```

Here again, ELF (but now also LZ150 and 256V16 is just 0.02% off for V) believe the 2nd-highest-prior move F2 should be played with the higher win rate if not for the policy. So why didn't ELF training favor increasing prior for F2 relative to what we have with leelaz?

Perhaps one explanation for the priors not adjusting to the higher win rate move is that there's just so much training data favoring the current policy, so even with noise, the average training filters out the noise. If that's the case, should something be done to have training's root search go wider (e.g., larger epsilon from 25% to 50% or even higher when applying noise)?

Although with more root moves being searched during training, T=1 will probably also make win rates converge slower for the root position. But then again, T=1 is the way these not-highest-prior moves can get their value to increase so future training search drives more visits to increase priors. The custom code to require more than 1 visit for T=1 to select probably helps quite a bit when value is already pretty good.

**Comment by Friday9i** [comment]
Rather than increasing noise (which is already high...), wouldn't it be useful to just make a wider tree exploration?
For that, we could simply modify the softmax calculation of moves' probabilities.
Practically, the nets gives an input for each move, and the probability is computed using softmax(input) for each move. Instead of using softmax(input), we could use softmax(a*input), with a=0.5 for example.
Exemple: if there were only 3 possible moves with inputs (4.5, 2, 0.3), softmax gives probabilities of 91%, 7.5% and 1,4% respectively, which is very focused on the first move. Hence, the tree-search will be focused on it and move 3 will almost be ignored. Using softmax(input/2), it gives probabilities of 71%, 20.3% and 8.7%, which is much more spread among the 3 moves, and the search will explore the 3 of them.
In the exemple you gives, that would easily allow the program to explore more moves and discover after some search that several of them are quite attractive ; -)
Currently, I get the impression that LZ (as well as AG/AGZ/AZ) are very focused on a few moves and are not exploring enough alternatives...: that approach could change things and LZ could explore many more josekis ; -).
What do you think of that?

**Comment by gcp** [comment]
I don't understand why the topic is called "prevent learning even with noise" and the discussion shows searches without noise.

There's no guarantee that with more time or larger network capacity the network wouldn't train exactly as you expect. Remember also that the networks do generalize their knowledge, so the score of priors isn't exclusively caused by what the network has learned for that position only.

>Rather than increasing noise (which is already high...), wouldn't it be useful to just make a wider tree exploration?
>For that, we could simply modify the softmax calculation of moves' probabilities.

All of this is already included as a parameter in the engine and has been forever.

But again, there's no reason whatsoever to even think this would be required. Everyone and their dog always wants to make suggestions to make the engine search wider because they find single positions where the engine doesn't find the best move, continuously ignoring that this has been shown time and time again to make it weaker.

**Comment by gcp** [comment]
>for sure in match games, where it should be disabled.
>But for training, isn't it complementary to noise, in order to explore and discover new moves more frequently and easily?

There's a trade-off between finding new moves and learning to predict more accurately. I would say it's not necessarily complimentary, but might just be redundant.

Also: if it would help self-play to discover new moves, why would it not help the regular search in match games to close holes in the network's knowledge?

>The simple fact that just a few josekis are played for weeks, then droped for a few new josekis, illustrates the fact that LZ is "specialized", I'd say a bit too specialized.

Does it not demonstrate the opposite: that the learning works? For the learning to work efficiently, the data must not be too noisy (or it'll be slower), and the MCTS gain must be high (so searching ahead teaches the bot new things) - which requires a strong search - which we know must not explore too much. Despite the focus, the engine still manages to discover new things and shift away.

>but a larger exploration at the time would probably have helped LZ to discover alternative josekis earlier

As explained above, exactly the opposite might also have happened.

You are right that more exploration may or may not help for learning (and in our setup there's no really good ways to test). 

But I am observing that the following sequence keeps happening again and again and again:
- Person sees a game where LZ loses or does not play the best move.
- Person loads the position and sees that with forcing a different exploration, the engine finds the best move or a better one.
- Person makes *any* suggestion that results in a different exploration order, or more exploration order, and thinks it may improve the engine.
- (Optional:) The idea is tested and just like all others before it that increase exploration, it turns out to make the engine weaker.

For the reasoning given above, this way of thinking is extremely and deeply misleading and can only lead to improvements "by accident".

**Comment by Friday9i** [comment]
"the opposite might also have happened": possible indeed, that's why "I'll do my best to try to test it by myself" (modifying the code and testing it, if I manage to do it : -).
Thanks for your detailed and clear comments.

**Comment by Mardak** [comment]
> Despite that a strong search focus is one of the major contributors to engine performance, people seldomly suggest things that make it more selective (FPU reductions being a nice one).

Your example of fpu reduction and the subsequent fix to remove it for noised self-play is almost proof that wider search is beneficial for learning.

One of the assumptions I made with this post is that ELF is trained sufficiently longer than what we have done so far with leelaz, and that implies noise has been used in training (hence the title). Specifically looking at the win rate and prior information:
```
LZ   F2 ->    1000 (V: 47.39%) (N: 26.11%)
LZ   E3 ->    1000 (V: 46.70%) (N: 71.04%)
ELF  F2 ->    1000 (V: 49.65%) (N: 22.23%)
ELF  E3 ->    1000 (V: 48.32%) (N: 77.73%)
```

Why do both networks prefer the low winrate move E3 by giving it a higher prior?

> Remember also that the networks do generalize their knowledge, so the score of priors isn't exclusively caused by what the network has learned for that position only.

Yes, I believe this is exactly what's causing LZChess to train away from the correct move for various tactics as a position where a queen sacrifice is good looks very much like positions where queen sacrifices are bad, and the latter bad positions are way more common than the former. A larger network capacity could indeed help that particular situation.

For both Go and Chess, this generalized knowledge aspect practically means more visits for the non-highest-prior move in self-play search are needed for training to overcome this same position's priors as well as similar/generalized position training data visit percentages.

>  For that, we could simply modify the softmax calculation of moves' probabilities.

As gcp pointed out, there is an existing cfg_softmax_temp, and I originally included it in the example of ways to make search wider, but I didn't want to complicate the original post further. At least for LZChess, because the tactical move has extremely low prior around 0.5%, adjusting the policy softmax even just a bit over 1 makes a huge difference. This has a much bigger impact than increasing noise epsilon from 25% to 50%. Also, complicating the softmax idea is that it would normally affect non-root moves, which we probably don't want for training, so that's why I only suggested adjusting existing noise value, which only affects root.

Basically, for self-play search, it should get an accurate estimation of win rate based on its learned priors for non-root searched moves so that root search can better direct visits for valuable training data. This is exactly how noise behaves now, but perhaps it's not enough at this stage of learning? One could expect DeepMind not wanting to have additional parameters to tune during a run other than learning rate, so it doesn't seem too unlikely that the default noise was selected for all phases of training (or at least for early network learning without significantly restricting later learning).

**Comment by Marcin1960** [comment]
@Friday9i "we could use softmax(a * input), with a=0.5 for example"

Can I make such or similar change in the code for leelaz.exe? If so, where should I do it?

**Comment by Eddh** [comment]
OpenAI is having a lot of success with more randomized training conditions right now https://blog.openai.com/openai-five/

It does support the idea that exploration in training is something important to take care of, and it is not simply enough that exploration is possible, it should also be sufficiently high. We would need multiple training runs to find out if we are currently not exploring enough, with softmax, temperature tuning etc.

The randomized training conditions that OpenAI use offer a possible other way to increase exploration in training. We could do the same with certain parameters, such as visits per move for each player, cpuct, or noise parameters. Currently the nets are trained only on 1 set of these parameters, so it expects the opponent to play at a very similar level as him. Randomized training condition could make it better at playing against stronger or weaker opponents, and explore different ways of playing the game.

The bad point would be the new overhead of having another set of parameter, the randomization ones, in LeelaZero, there is already -m and -n but this would add new ones which might be confusing. On the other hand such a change would not be very hard to implement.

What do you think ?

**Comment by Mardak** [comment]
> There's an incentive in a learning setup for learning to make sure things are never permanently discarded, which is why LZ/AGZ has the mechanics that are there.

Just to be clear, what are the mechanics you're referring to? If it's applying Dirichlet noise with the alpha and epsilon, that's entirely the premise of this issue questioning if the current α=3% and ε=25% are sufficient to make sure moves are never permanently discarded or prevented from learning to play the best (highest value) move.

**Comment by gcp** [comment]
> At least for LZChess, because the tactical move has extremely low prior around 0.5%, adjusting the policy softmax even just a bit over 1 makes a huge difference. 

So, does changing the policy softmax to max(softmax, 0.01) actually make LZC stronger?

It not, please reread: https://github.com/gcp/leela-zero/issues/1588#issuecomment-400278489

**Comment by Mardak** [comment]
> Pruning moves in such a way they can never be reconsidered by the learning process is probably a bad idea. This is quite far away from "search wider".

Why was fpu_reduction removed from noised root? The before and after effect is that self-play searched wider when no longer applying fpu_reduction. Pruning moves = narrower search, not pruning = wider. I'm not suggesting widest search is desirable, just that there has been a change to make it wider. Should search be wider than it is now? Unclear.

> The only thing you need is a positive feedback cycle where the prior can increase network over network, you don't need to "overcome" anything.

A single training data of favorably noised play produces positive / increased prior training data, but that doesn't make a positive feedback cycle as when it's averaged with unfavorably noised data as well as those from similar/generalized positions, that potentially increased prior is lost.

> > Also, complicating the softmax idea is that it would normally affect non-root moves, which we probably don't want for training

> The logic here completely escapes me. Why wouldn't you want this to be generic, if it actually helps?

Generic meaning also use for match games? The purpose of training priors is to make search more efficient focusing only on likely moves, and the efficiency is more important in matches when there's a time aspect.

Or generic meaning for use in non-root self-play moves? Having softmax-adjusted priors for non-root moves perturbs the win rate that search is trying to maximize.

**Comment by Mardak** [comment]
> So, does changing the policy softmax to max(softmax, 0.01) actually make LZC stronger?

I don't think that specific behavior has been tested, but a similar behavior of changing policy with softmax temp 2.2 has been measured to improve match performance:
https://github.com/LeelaChessZero/lc0/pull/46

For reference, here's the prior policy for some tactics positions that LZC fails to find without softmax but does with softmax temp 2.2:
```
   sctr Rxh4 0.11% -> 1.15%
   wasp Rxe3 0.26% -> 1.51%
   naum Qxd5 0.16% -> 1.32%
vajolet Nxg4 0.15% -> 1.19%
```

The adjusted policy is not too far from your suggested 1%.

So increasing policy for low moves can be beneficial for matches, but I believe the more appropriate place to search wider is during training. Study before a test and stick with the learned behavior instead of trying things on-the-fly. (Unless of course, the network just doesn't have the capacity to learn and needs to be compensated with adjusting policy.)

**Comment by gcp** [comment]
> a similar behavior of changing policy with softmax temp 2.2 has been measured to improve match performance

lc0 is a very bad thing to do these kind of measurements on because it requires a hugely wide search to get efficient batching, which is a completely orthogonal thing from self-play games and training.

I believe Facebook has said the same of ELF, which also uses different parameters for self-play vs matches there.

**Comment by alreadydone** [comment]
What parameters are different? At least c_puct is the same: https://github.com/pytorch/ELF/issues/53

**Comment by Mardak** [comment]
> lc0 is a very bad thing to do these kind of measurements on because it requires a hugely wide search to get efficient batching in matches

I'm pretty sure ASilver did the tuning run with lczero, but it happened to only be implemented with lc0 (with cpuct scaled appropriately).

**Comment by gcp** [comment]
>What parameters are different? At least c_puct is the same: pytorch/ELF#53

Ah, it seems that they cleared this up. I remembered the values being different.

**Comment by gcp** [comment]
>I'm pretty sure ASilver did the tuning run with lczer

```
The settings are only valid for LC0 not LCZero.

PUCT: 3.4
FPU: 0.9
Policy: 2.2
```

**Comment by gcp** [comment]
FWIW I never tried tuning cfg_softmax_temp because it makes the mapping from the probabilities produced by the search into the desired network outputs rather weird. DeepMind had it at ~0.67 for Alpha Lee Go but fixed it at 1.0 for A(G)Z.

**Comment by Mardak** [comment]
> It will go upwards when the move score is good enough.

That's what I've been trying to highlight with the example position with latest LZ and ELF weights. Both networks could find a non-highest-prior move with a better win rate (roughly 1 percentage point higher) but training data seems to reinforce putting visits into the lower win rate move.

>> is more important in matches when there's a time aspect.

> There is always a time aspect. Faster self-play will cause faster improvement.

There are two different time scales that we shouldn't conflate. One is for finding the correct move during a game. Another is time to improve the network (through generating self-play, training, promoting, etc.). Also, the speed of self-play is fixed with visits anyway (but there's separate discussions on reducing visits).

Optimizing the former time affects how strong a network actually performs whereas the latter of network improving seems to be more flexible.

>> Having softmax-adjusted priors for non-root moves perturbs the win rate that search is trying to maximize.

> So is any method that searches wider?

Again, don't want to conflate two different win rates. Search is maximizing win rates to pick the best move with value and policy trying to reflect what the network believes strong play will be (but yes, the network could be mistaken about what the win rate and plays should be). And yes, searching wider during training along with T=1 will make the learned value converge slower. 

> (It causes a slower convergence from mean -> max in the node value)

Although specifically to this concern, search doesn't use the root's current eval anyway? Resignation is based on the most visited move as well. So slower convergence at root shouldn't affect training much (or at all??)?

> The settings are only valid for LC0 not LCZero.

I was mistaken and did confirm with ASilver that the tuning was with lc0. It was with id369 done at a time when code was moving to its own repository and ongoing development was shifting from lczero to lc0. The example prior adjusting I provided earlier were with lczero though using `--softmax_temp`, and search does indeed find the correct move when policies increase from 0.1% to 1.0%.

**Comment by gcp** [comment]
>training data seems to reinforce putting visits into the lower win rate move

I don't see how you can conclude this. The examples you gave had visits fixed at 1000 for every move. If it were not, then the visits would be strongly proportional to the win rate and to a lesser extent proportional to the prior.

A move with a high winrate would get a lot of visits and feed back to learn a higher prior.

```
  F2 ->    1000 (V: 47.39%) (N: 26.11%) PV: F2 G2 E2 G4 E4 E3 B4 B5 B3 C6 E5 D6 E6 E7 F7 F6 E8 D7 G6 F5 G5 F4 H4 G3 G7 R3 R4 Q3 P4 O2 R16 R17 S17 R15 S16 Q15 S15 R18 Q14
  E3 ->    1000 (V: 46.70%) (N: 71.04%) PV: E3 E4 F2 R3 Q3 R4 R6 R5 Q5 S6 G3 F4 B3 H5 R2 S2 S7 Q2 Q7 P6 Q6 P2 F17 C14
```

If visits were distributed according to UCT (as in self-play), there's a good likelihood it would be training to eventually prefer F2. If discovering that F2 is a bit border-line, it might not reach that point with few visits until first learning more about the successor positions.

>Also, the speed of self-play is fixed with visits anyway

You can lower visits if the gain from the MCTS is higher (or if you drop gating, but again, different discussion).

>So slower convergence at root shouldn't affect training much (or at all??)?

I was responding to a statement of yours about non root moves: ```Having softmax-adjusted priors for non-root moves...```

>search does indeed find the correct move when policies increase from 0.1% to 1.0%.

For the third time: https://github.com/gcp/leela-zero/issues/1588#issuecomment-400278489

**Comment by Mardak** [comment]
> If visits were distributed according to UCT (as in self-play), there's a good likelihood it would be training to eventually prefer F2.

I ran the analysis forcing 1000 visits to each of the moves to have comparable win rates given same amount of searching. I too would expect the network to eventually prefer F2, but looking at LZ now and ELF, which is roughly LZ "eventually," it shows that ELF still believes F2 has the higher win rate but its prior has not trained to reflect that.

**Comment by Mardak** [comment]
> I was responding to a statement of yours about non-root moves: `Having softmax-adjusted priors for non-root moves...`

I think we're in agreement here then? You're pointing out that a child's eval will converge slower with wider search, and I'm saying we don't want softmax policy for non-root because we want an accurate measurement of win rate for a move.

So that's why noise / wider search / flatter policy should only be applied at the root and not for children. (I brought this up originally because a simple setting of cfg_softmax_temp will adjust the policy for all nodes, so additional code would be needed to make it work only at the root.)

**Comment by Nazgand** [comment]
I recall that the later version, AlphaZero, does not have a policy network separate from the value network. A single network was used for both purposes.

**Comment by gjm11** [comment]
Leela Zero also uses a single network for both purposes.

**Comment by Nazgand** [comment]
Perhaps this problem could be solved by training the move suggestion of the policy network of the current position to match the win-rate of the value network after the move.

Would forcing a playout of every legal move help to solve the problem? @Mardak I am not sure exactly how the tree search works, yet I suspect having a single playout with a high win-rate would bias the tree search towards good moves even when the policy network is biased against them.

**Comment by sethtroisi** [comment]
I think recent minigo runs have shown yes.
